rewrite-graphql/src/main/java/org/openrewrite/graphql/internal/GraphQlPrinter.java

```diff
@@ -0,0 +1,1083 @@
+        beforeSyntax(fieldDef, p);
+        if (fieldDef.getDescription() != null) {
+            visitStringValue(fieldDef.getDescription(), p);
+            p.append(" ");
+        }
+        visitName(fieldDef.getName(), p);
+        visitInputValueDefinitions(fieldDef.getArguments(), fieldDef.getArgumentsEnd(), p);
+        p.append(":");
+        visitType(fieldDef.getType(), p);
+        visitList(fieldDef.getDirectives(), p);
+        afterSyntax(fieldDef, p);
+        return fieldDef;
+    }
+
+    @Override
+    public GraphQl visitInputValueDefinition(GraphQl.InputValueDefinition inputValueDef, PrintOutputCapture<P> p) {
+        beforeSyntax(inputValueDef, p);
+        if (inputValueDef.getDescription() != null) {
+            visitStringValue(inputValueDef.getDescription(), p);
+            p.append(" ");
+        }
+        visitName(inputValueDef.getName(), p);
+        p.append(":");
+        visitType(inputValueDef.getType(), p);
+        if (inputValueDef.getDefaultValue() != null) {
+            p.append(" =");
+            visitValue(inputValueDef.getDefaultValue(), p);
+        }
+        visitList(inputValueDef.getDirectives(), p);
+        afterSyntax(inputValueDef, p);
+        return inputValueDef;
+    }
+
+    @Override
+    public GraphQl visitInterfaceTypeDefinition(GraphQl.InterfaceTypeDefinition interfaceType, PrintOutputCapture<P> p) {
+        beforeSyntax(interfaceType, p);
+        if (interfaceType.getDescription() != null) {
+            visitStringValue(interfaceType.getDescription(), p);
+            p.append(" ");
+        }
+        p.append("interface");
+        visitName(interfaceType.getName(), p);
+        visitImplementsInterfaces(interfaceType.getImplementsInterfaces(), p);
+        visitList(interfaceType.getDirectives(), p);
+        if (interfaceType.getFieldsPrefix() != null) {
+            printSpace(interfaceType.getFieldsPrefix(), p);
+            if (interfaceType.getFields() != null && !interfaceType.getFields().isEmpty()) {
+                visitContainer("{", interfaceType.getFields(), "}", null, p);
+            } else {
+                p.append("{}");
+            }
+        }
+        afterSyntax(interfaceType, p);
+        return interfaceType;
+    }
+
+    @Override
+    public GraphQl visitUnionTypeDefinition(GraphQl.UnionTypeDefinition unionType, PrintOutputCapture<P> p) {
+        beforeSyntax(unionType, p);
+        if (unionType.getDescription() != null) {
+            visitStringValue(unionType.getDescription(), p);
+            p.append(" ");
+        }
+        p.append("union");
+        visitName(unionType.getName(), p);
+        visitList(unionType.getDirectives(), p);
+        if (unionType.getMemberTypes() != null && !unionType.getMemberTypes().isEmpty()) {
+            p.append(" =");
+            for (int i = 0; i < unionType.getMemberTypes().size(); i++) {
+                if (i > 0) {
+                    p.append(" |");
```

Comment:
Printing single space literals like this is always fishy. Probably means what we should look into making something a `LeftPadded` or `RightPadded` somewhere.